trio/_core/_result.py

```diff
@@ -1,125 +1,29 @@
-    """An abstract class representing the result of a Python computation.
-
-    This class has two concrete subclasses: :class:`Value` representing a
-    value, and :class:`Error` representing an exception.
-
-    In addition to the methods described below, comparison operators on
-    :class:`Value` and :class:`Error` objects (``==``, ``<``, etc.) check that
-    the other object is also a :class:`Value` or :class:`Error` object
-    respectively, and then compare the contained objects.
-
-    :class:`Result` objects are hashable if the contained objects are
-    hashable.
-
-    """
-    __slots__ = ()
-
-    @staticmethod
-    def capture(sync_fn, *args):
-        """Run ``sync_fn(*args)`` and capture the result.
-
-        Returns:
-          Either a :class:`Value` or :class:`Error` as appropriate.
-
-        """
-        try:
-            return Value(sync_fn(*args))
-        except BaseException as exc:
-            return Error(exc)
-
-    @staticmethod
-    async def acapture(async_fn, *args):
-        """Run ``await async_fn(*args)`` and capture the result.
-
-        Returns:
-          Either a :class:`Value` or :class:`Error` as appropriate.
-
-        """
-        try:
-            return Value(await async_fn(*args))
-        except BaseException as exc:
-            return Error(exc)
-
-    @abc.abstractmethod
-    def unwrap(self):
-        """Return or raise the contained value or exception.
-
-        These two lines of code are equivalent::
-
-           x = fn(*args)
-           x = Result.capture(fn, *args).unwrap()
-
-        """
-
-    @abc.abstractmethod
-    def send(self, gen):
-        """Send or throw the contained value or exception into the given
-        generator object.
-
-        Args:
-          gen: A generator object supporting ``.send()`` and ``.throw()``
-              methods.
-
-        """
-
-    @abc.abstractmethod
-    async def asend(self, agen):
-        """Send or throw the contained value or exception into the given async
-        generator object.
-
-        Args:
-          agen: An async generator object supporting ``.asend()`` and
-              ``.athrow()`` methods.
-
-        """
-
-
-@attr.s(frozen=True, repr=False)
-class Value(Result):
-    """Concrete :class:`Result` subclass representing a regular value.
-
-    """
-
-    value = attr.ib()
-    """The contained value."""
-
-    def __repr__(self):
-        return "Value({!r})".format(self.value)
-
-    def unwrap(self):
-        return self.value
-
-    def send(self, gen):
-        return gen.send(self.value)
-
-    async def asend(self, agen):
-        return await agen.asend(self.value)
+from .._deprecate import deprecated
 
+__all__ = ["Result", "Value", "Error"]
 
-@attr.s(frozen=True, repr=False)
-class Error(Result):
-    """Concrete :class:`Result` subclass representing a raised exception.
 
-    """
+class Result(outcome.Outcome):
+    @deprecated(version="0.5.0", issue=494, instead="outcome.Outcome")
```

Comment:
`Result` was previously an ABC, so I wouldn't expect anyone to be trying to construct it; I think the deprecation notice would need to go on capture(), acapture(), and the constructors of Value and Error in order to have any effect.

Reply:
For `Value` and `Error` I'd suggest using `trio._deprecate`'s trick for module level deprecations. It lets you say "when someone accesses `trio.hazmat.Value`, then give them [this] deprecation message and then return [this] object." It looks like the diff for 089c6ff02e624190d4cd3c4f3f2543cabcddfdec has an example.

It might even be simplest to use this for `Result` too. The advantage is that it gives a warning for `isinstance(x, trio.hazmat.Result)`.